File: services/api/src/systems/message-queue/rabbit-mq-publisher.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { RabbitMQConnectionService } from './rabbit-mq-connection.service';

@Injectable()
export class RabbitMQPublisherService {
  private readonly logger = new Logger(RabbitMQPublisherService.name);

  constructor(
    private readonly rabbitMQConnectionService: RabbitMQConnectionService,
  ) {}

  async addMessageToQueue(queue: string, messageData: unknown) {
    const channel = this.rabbitMQConnectionService.getChannel();
    await this.rabbitMQConnectionService.assertQueue(queue);
    channel.sendToQueue(queue, this.convertData(messageData), {
      persistent: true,
    });

    this.logger.debug(
      `Message sent to queue (${queue}) with task data ${messageData}`,
    );
  }

  convertData(data: unknown) {
    return Buffer.from(JSON.stringify(data));
  }
}

File: services/diseases-analyzers/app/services/rabbitmq_publisher_service.py
import json
import logging
import pika
from app.services.rabbitmq_connection_service import RabbitMQConnectionService

class RabbitMQPublisherService:
    def __init__(self, rabbitmq_connection_service: RabbitMQConnectionService):
        self.logger = logging.getLogger(__name__)
        self.rabbitmq_connection_service = rabbitmq_connection_service

    def add_message_to_queue(self, queue: str, message_data: dict):
        channel = self.rabbitmq_connection_service.get_channel()
        self.rabbitmq_connection_service.assert_queue(queue)
        self.logger.debug(f"Sending message to queue: {queue}. Data: {message_data}")
        channel.basic_publish(
            exchange='',
            routing_key=queue,
            body=self.convert_data(message_data),
            properties=pika.BasicProperties(delivery_mode=2)
        )
        print(f"Message sent to queue ({queue}) with task data {message_data}")

    def convert_data(self, data: dict) -> bytes:
        return json.dumps(data).encode('utf-8')
